pkg/flow/events.go

```diff
@@ -1005,3 +1015,175 @@ func (events *events) listenForEvents(ctx context.Context, im *instanceMemory, c
+		if err != nil {
+			return resp, err
+		}
+
+		script = ceventfilter.Jscode
+
+	}
+
+	var mapEvent map[string]interface{}
+	err = json.Unmarshal(cloudevent, &mapEvent)
+
+	if err != nil {
+		return resp, err
+	}
+
+	//create js runtime
+	vm := goja.New()
+
+	vm.Set("event", mapEvent)
+
+	_, err = vm.RunString(script)
+	if err != nil {
+		//panic(err)
+		return resp, err
+	}
+
+	var fn func() map[string]interface{}
+	err = vm.ExportTo(vm.Get("filter"), &fn)
+	if err != nil {
+		//panic(err)
+		return resp, err
+	}
+
+	newEventMap := fn()
+	newBytesEvent, err := json.Marshal(newEventMap)
+	if err != nil {
+		return resp, err
+	}
+
+	resp.Event = newBytesEvent
+
+	if string(resp.GetEvent()) == "null" {
+		event, _ := EventByteToCloudevent(cloudevent)
+		if err != nil {
+			return resp, err
+		}
+		flow.logToNamespace(ctx, time.Now(), ns, "Dropped Event: %s", event.ID())
+	} else {
+		event, _ := EventByteToCloudevent(newBytesEvent)
+		if err != nil {
+			return resp, err
+		}
+		flow.logToNamespace(ctx, time.Now(), ns, "Eventfilter applied, new Event : %s (%s / %s)", event.ID(), event.Type(), event.Source())
+	}
+
+	return resp, err
+}
+
+func (flow *flow) DeleteCloudEventFilter(ctx context.Context, in *grpc.DeleteCloudEventFilterRequest) (*emptypb.Empty, error) {
+
+	var resp emptypb.Empty
+
+	namespace := in.GetNamespace()
+	filtername := in.GetFilterName()
+
+	ns, err := flow.getNamespace(ctx, flow.db.Namespace, namespace)
+	if err != nil {
+		return &resp, err
+	}
+
+	_, err = ns.QueryCloudeventfilters().Where(enteventsfilter.NameEQ(filtername)).Only(ctx)
+	if err != nil {
+		err = errors.New("event filter not existing")
+		return &resp, err
+	}
+
+	_, err = flow.db.CloudEventFilters.
+		Delete().
+		Where(
+			enteventsfilter.And(
+				enteventsfilter.NameEQ(filtername),
+			)).
+		Exec(context.Background())
+
+	if err != nil {
+		return &resp, err
+	}
+
+	var key keyPair
+	key.filtername = filtername
+	key.namespace = namespace
+	delete(cache, key)
+
+	return &resp, err
+
+}
+
+func (flow *flow) CreateCloudEventFilter(ctx context.Context, in *grpc.CreateCloudEventFilterRequest) (*emptypb.Empty, error) {
+
+	var resp emptypb.Empty
+
+	namespace := in.GetNamespace()
+	filtername := in.GetFiltername()
+	script := in.GetJsCode()
+
+	script = fmt.Sprintf("function filter() {\n %s \n}", script)
+
+	//compiling js code is needed
+	_, err := goja.Compile("filter", script, false)
+	if err != nil {
+		err := errors.New("js code does not compile")
+		return &resp, err
+	}
+
+	ns, err := flow.getNamespace(ctx, flow.db.Namespace, namespace)
+	if err != nil {
+		return &resp, err
+	}
+
+	ceventfilter, _ := ns.QueryCloudeventfilters().Where(enteventsfilter.NameEQ(filtername)).Only(ctx)
+	if ceventfilter != nil {
+		err = errors.New("event filter already exists")
+		return &resp, err
+	}
+
+	_, err = flow.db.CloudEventFilters.Create().SetName(filtername).SetNamespace(ns).SetJscode(script).Save(context.Background())
```

Comment:
When you already have a context (arg 0 `ctx`), it's *usually* not correct to use the background context.